### ISSUE TITLE
s.queueOut: timeout

### ISSUE THREAD

**Issue by afoninsky** [tinode/chat#245]
Hello.

### Question:
I have lots of `queueOut` timeouts during the process of developing custom grpc client. I create a new session and start to lose messages right from the beginning.

According the following code:
https://github.com/tinode/chat/blob/eef5b5cca2a34f5b6d9127af27a69b47a9f762d0/server/sessionstore.go#L63
https://github.com/tinode/chat/blob/f70f14dad5a2217d691021c6100026a80f4932b3/server/session.go#L177-L190

1) Why we create an output queue from 256 elements per session only - is it enough? Any message goes thru this queue, including presence and so , kinda lot of traffic.
2) Can we increase the timeout of 50ms? I think it is OK to wait much more.
3) I'm not sure that discarding messages is a good idea - this is a serious bug which lead to implicit behaviour. It is better either wait until last or close the session (allowing to reconnect by the client). Instead, I receive {data} but not {ctrl}, etc. It definitely breaks API flow.

Please fix me if I wrong. Thank you.

### My environment:
- tinode/tinode-rethinkdb:0.15.14
- GRPC client

### Server-side log example:
```
2019/06/18 06:44:21 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:44:21 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:45:30 grpc in: sub:<id:"66F30D03-E935-4D15-85AB-FC8D93C109BB" topic:"new" set_query:<desc:<public:"{\"title\":\"usrVvrZf1sFU9c\",\"fn\":\"SunovaMax\"}" > > get_query:<what:"desc" > >  VTeQsQWMEgY
2019/06/18 06:45:32 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:45:32 grpc in: set:<id:"2359E3DE-1309-41FF-A48C-7AB0B80FF43F" topic:"grpgoBdfcEsvi0" query:<desc:<public:"{\"title\":\"voiceChatDev7:1560840330.814188_usr9XFrl4BQWfM\"}" > sub:<user_id:"usr9XFrl4BQWfM" mode:"JRWPSA" > > >  VTeQsQWMEgY
2019/06/18 06:45:32 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:45:32 grpc in: sub:<id:"D37FCAE7-23F5-4310-A65A-093BF5D35F79" topic:"grpgoBdfcEsvi0" get_query:<what:"desc" > >  hywcSg77xMs
2019/06/18 06:45:32 grpc in: pub:<id:"1B814416-18F4-4421-8AD5-9918D3901D9C" topic:"grpgoBdfcEsvi0" content:"\"voiceChatDev7:connected\"" >  hywcSg77xMs
2019/06/18 06:45:33 grpc in: pub:<id:"D107E785-4474-490F-831E-016980A76F6E" topic:"grpgoBdfcEsvi0" content:"\"voiceChatDev7:callusr9XFrl4BQWfM\"" >  VTeQsQWMEgY
2019/06/18 06:45:33 grpc in: pub:<id:"E6BB048F-739C-4A64-AB37-E056AB0A3926" topic:"grpgoBdfcEsvi0" content:"\"voiceChatDev7:decline\"" >  hywcSg77xMs
```

**Comment by or-else** [comment]
1. Something is fishy with the client. Why isn't it keeping up? 256 is a lot. It's not hard to change the size of the queue but (a) I would try to understand how come the queue is full, (b) too large of a queue may allow malicious clients to waste server memory. 
This is a queue per connection. I.e. 256 messages were queued to be sent to a single client. Seems like the client is not reading.

2. It can be changed but I don't think it's the source of the problem.

3. If not discarding then what else should the servers do? It cannot queue indefinitely. There must be a limit on the number of queued messages. So, when the limit is reached it can either start dropping messages or drop the entire connection. Dropping the connection is probably more logical. 

I would first try to understand why the messages are being queued so much.

**Comment by afoninsky** [comment]
You're right - increasing buffer size won't help. I've added debug messages and see that timeout fires when queue is empty or almost empty.

But also I'm not sure it is a client problem (or at least it should be a complex one). Because I have the same behaviour on fresh tinode instances and default web client.

For example, I login with the default "alice" and start sending messages. As you see one message is not acknowledged and there is a `s.queueOut: timeout FDXHpRkoJr0` message in the log.

![image](https://user-images.githubusercontent.com/15193941/59849640-e343c380-9370-11e9-9845-799f43de1cda.png)

Feel free to test it by yourself, it is easy to reproduce: http://35.237.57.206:18080/
Again, everything is default and I haven't changed nothing.

I continue researching but don't mind about some advices because have no idea where to dig so far :)

My env:
kubernetes v1.13.6
image tinode/tinode-rethinkdb:0.15.14
RethinkDB 2.3.6
config: https://gist.github.com/afoninsky/5b95d66138e83fe0c83756f2c433f4eb

**Comment by or-else** [comment]
I don't see this at all at `api.tinode.co`:

```
user@tinode-small:~$ cat /var/log/tinode.log | wc -l
246
user@tinode-small:~$ cat /var/log/tinode.log | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.0 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.1 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.2 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.3 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.4 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.5 | grep timeout
user@tinode-small:~$ cat /var/log/tinode.log.6 | grep timeout
user@tinode-small:~$ 
```

Not a single timeout in a week worth of logs. It's probably something related to your setup.

Can you add debug printing of `len(s.send)` at https://github.com/tinode/chat/blob/master/server/session.go#L202 like 
```
log.Println("s.queueOutBytes: timeout", s.sid, len(s.send))
```

As a test you can try to increase the timeout from 50usec to maybe 100usec, but it's still very odd.

**Comment by afoninsky** [comment]
I did both actually: increased timeouts, queue size and added logs. Increasing does not help. Logs say there are 0 or 1 message in the queue.
As I understand - `select` will block pushing to the channel either if channel buffer is full or there are no one can read from the channel. The second case is ours - but I have no idea why. Maybe client is not able to receive it? Dunno.

I was able to repeat this behaviour on two different kubernetes clusters (GKE) with various clients (websocket, grpc). And I have not seen same issue руку - so I agree ерфе it is environment specifiс stuff.

Will continue research :)

**Comment by or-else** [comment]
> The second case is ours - but I have no idea why. Maybe client is not able to receive it?

That's not the remote client. It's this one in case of grpc:
https://github.com/tinode/chat/blob/master/server/hdl_grpc.go#L76

or this in case of a websocket:
https://github.com/tinode/chat/blob/master/server/hdl_websock.go#L77

Maybe this call blocks in your case?
https://github.com/tinode/chat/blob/master/server/hdl_grpc.go#L102
Please wrap it in a timer and if it's more than 1ms print out a log record.

Websocket write has a deadline. gRPC write does not.

**Comment by or-else** [comment]
It looks like once the timeout happens, the subsequent sends always timeout:
```
2019/06/18 06:44:21 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:44:21 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:45:30 grpc in: sub:<id:"66F30D03-E935-4D15-85AB-FC8D93C109BB" topic:"new" set_query:<desc:<public:"{\"title\":\"usrVvrZf1sFU9c\",\"fn\":\"SunovaMax\"}" > > get_query:<what:"desc" > >  VTeQsQWMEgY
2019/06/18 06:45:32 s.queueOut: timeout VTeQsQWMEgY
2019/06/18 06:45:32 grpc in: set:<id:"2359E3DE-1309-41FF-A48C-7AB0B80FF43F" topic:"grpgoBdfcEsvi0" query:<desc:<public:"{\"title\":\"voiceChatDev7:1560840330.814188_usr9XFrl4BQWfM\"}" > sub:<user_id:"usr9XFrl4BQWfM" mode:"JRWPSA" > > >  VTeQsQWMEgY
2019/06/18 06:45:32 s.queueOut: timeout VTeQsQWMEgY
```

That's consistent with `s.send` channel being `nil`. Could you please check if `s.send` is `nil` when the timeout happens?

**Comment by or-else** [comment]
Are you running a modified server?

**Comment by afoninsky** [comment]
> Are you running a modified server?
No.

I didn't find the root of issue, but found the way how to avoid it. In general - do not use GKE :)

Following setup does not work: "[google load balancer]->[kuberbetes service]->[kubernetes tinode docker]"
But this one works: "[google load balancer]->[kubernetes service]->[tinode docker on debian outside of kubernetes]"

Still have no idea what goes wrong - either I was the first one who met this problem (because I was able to repeat it on different GKE clusters) or it's only my failure. My last idea was about complex bug between outgress traffic routing implementation in GKE and tinode sending queue implementation but I didn't test it.


As I get working setup - going to stop research for now and will continue later or/and wait for similar issues.
I can create test stand if you willing to dig or try to perform actions you would ask.

**Comment by or-else** [comment]
> I can create test stand if you willing to dig

Sure. I would give it a shot. Thanks.

**Comment by or-else** [comment]
Is it possible that kubernetes pod with tinode server is preempted while waiting on `select`?

**Comment by afoninsky** [comment]
What do you mean? Throttling? It makes sense.
For now, I can't see any significant changes between GCE virtual instance and GKE host node where tinode launches as a pod. Except OS and etc.

> Sure. I would give it a shot. Thanks.
Great, can you give me your gmail email? I will deploy test cluster and setup permissions to access so we can debug together.

**Comment by or-else** [comment]
If I understand resource allocation in kubernetes correctly, an instance gets a slice of a CPU time, like 

```
resources:
  requests:
    memory: 50Mi
    cpu: 50m
  limits:
    memory: 100Mi
    cpu: 100m
```
an instance in this example gets 10% of CPU time. Is it possible that the instance is suspended for more than 50 usec to enforce the CPU limit?

I'm a complete nubie with respect to kubernetes.

**Comment by or-else** [comment]
I'm at gene@tinode.co

**Comment by afoninsky** [comment]
> I'm at gene@tinode.co

Thank you for your support. I think I've found an issue.

> Is it possible that the instance is suspended for more than 50 usec to enforce the CPU limit?

Yes. But this is not quite CPU limit trouble. It's more about architecture problem described [here](https://stackoverflow.com/questions/54099425/pod-cpu-throttling).
In general: no mater how many resources you have allowed for your docker - only amount of resources consumed by siblings dockers matters.

I've created new kubernetes node and scheduled only tinode instance there using [tolerations](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) - everything is perfect.

So we have to keep in mind this restriction. Also increasing of 50 usec will slightly improve performance in such cases - I guess it's random one?
